src/__tests__/extensions/surveys.test.ts

```diff
@@ -31,28 +40,324 @@ describe('survey display logic', () => {
+        document.body.appendChild(surveyDiv)
+
+        expect(surveyManager.getTestAPI().canShowNextEventBasedSurvey()).toBe(true)
+
+        surveyDiv.shadowRoot!.appendChild(document.createElement('div'))
+        expect(surveyManager.getTestAPI().canShowNextEventBasedSurvey()).toBe(false)
+    })
+
+    test('callSurveysAndEvaluateDisplayLogic should handle surveys correctly', () => {
+        mockPostHog.getActiveMatchingSurveys = jest.fn((callback) => callback([mockSurveys[0]]))
+
+        const handlePopoverSurveyMock = jest
+            .spyOn(surveyManager as any, 'handlePopoverSurvey')
+            .mockImplementation(() => {
+                console.log('Mock handlePopoverSurvey called')
+            })
+        const handleWidgetMock = jest.spyOn(surveyManager as any, 'handleWidget').mockImplementation(() => {})
+        const handleWidgetSelectorMock = jest
+            .spyOn(surveyManager as any, 'handleWidgetSelector')
+            .mockImplementation(() => {})
+        jest.spyOn(surveyManager as any, 'canShowNextEventBasedSurvey').mockReturnValue(true)
+
+        surveyManager.callSurveysAndEvaluateDisplayLogic()
+
+        expect(mockPostHog.getActiveMatchingSurveys).toHaveBeenCalled()
+        expect(handlePopoverSurveyMock).toHaveBeenCalledWith(mockSurveys[0])
+        expect(handleWidgetMock).not.toHaveBeenCalled()
+        expect(handleWidgetSelectorMock).not.toHaveBeenCalled()
+    })
+
+    test('handleWidget should render the widget correctly', () => {
+        const mockSurvey = mockSurveys[1]
+        const handleWidgetMock = jest.spyOn(surveyManager as any, 'handleWidget').mockImplementation(() => {})
+        surveyManager.getTestAPI().handleWidget(mockSurvey)
+        expect(handleWidgetMock).toHaveBeenCalledWith(mockSurvey)
+    })
+
+    test('handleWidgetSelector should set up the widget selector correctly', () => {
+        const mockSurvey = {
+            id: 'widgetSurvey2',
+            name: 'Widget Survey 2',
+            type: SurveyType.Widget,
+            appearance: { widgetType: 'selector', widgetSelector: '.widget-selector' },
+            start_date: '2021-01-01T00:00:00.000Z',
+            questions: [
+                {
+                    question: 'What would you like to see next?',
+                    type: SurveyQuestionType.Open,
+                },
+            ],
+            conditions: {},
+            end_date: null,
+            targeting_flag_key: null,
+        } as any
+        document.body.innerHTML = '<div class="widget-selector"></div>'
+        const handleWidgetSelectorMock = jest
+            .spyOn(surveyManager as any, 'handleWidgetSelector')
+            .mockImplementation(() => {})
+        surveyManager.getTestAPI().handleWidgetSelector(mockSurvey)
+        expect(handleWidgetSelectorMock).toHaveBeenCalledWith(mockSurvey)
+    })
+
+    test('callSurveysAndEvaluateDisplayLogic should not call surveys in focus', () => {
```

Comment:
would love to see some tests around:

1.  The sorting behaviour for delayed + non-delayed surveys (and wth add an api survey in the mix too :P )
2. SurveyinFocus handling for `callSurveysAndEvaluateDisplayLogic` - i.e. assume there is a surveyInFocus, (like in this test) - then `handleSurveyPopover` is not called. Then you manually call `removeSurveyFromFocus` and then call `callSurveysAndEvaluateDisplayLogic` again and this time `handleSurveyPopover` is called.

3. [For a follow up]. - I don't remember if this already exists, but would love to see some tests around the `getSurveySeenKey` and local storage handling as well, but it's not something you're currently touching, so happy to defer to you whenever / if you want to add these

Reply:
I did the first 2, and then I'll do the next bit in my refactor PR.